cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/CompositeTrigger.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.schedule.trigger;
+
+import co.cask.cdap.internal.app.runtime.schedule.store.Schedulers;
+import co.cask.cdap.internal.schedule.trigger.ScheduleTrigger;
+import co.cask.cdap.internal.schedule.trigger.Trigger;
+import co.cask.cdap.proto.Notification;
+import co.cask.cdap.proto.ProtoTrigger;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Abstract base class for composite trigger.
+ */
+public abstract class CompositeTrigger extends ProtoTrigger.AbstractCompositeTrigger
+  implements SatisfiableTrigger {
+  protected boolean satisfied;
+  // A map of non-composite trigger type and set of triggers of the same type
+  private Map<Type, Set<Trigger>> unitTriggers;
+
+  public CompositeTrigger(Type type, Trigger... triggers) {
+    super(type, triggers);
+    initializeUnitTriggers();
+  }
+
+  public void initializeUnitTriggers() {
+    unitTriggers = new HashMap<>();
+    for (Trigger trigger : triggers) {
+      // Add current non-composite trigger to the corresponding set in the map
+      Type triggerType = ((ProtoTrigger) trigger).getType();
+      if (!triggerType.isComposite()) {
+        Set<Trigger> triggerList = unitTriggers.get(triggerType);
+        if (triggerList == null) {
+          triggerList = new HashSet<>();
+          unitTriggers.put(triggerType, triggerList);
+        }
+        triggerList.add(trigger);
```

Comment:
What is the point of this line? Don't you just add the trigger to the set, only for it to be redefined in the next iteration of this loop?